### ISSUE TITLE
How to get stream operators in custom backend compiler ?

### ISSUE THREAD

**Issue by wbigat** [pytorch/pytorch#130137]
### 🐛 Describe the bug

Hi, when I use a custom backend, I find that the fx graph that custom compiler gets does not have the stream related operations.

Then I found that the fx graph dropped those stream operations after aot_module_simplified.
So, I want to know how can we get a fx graph that contains stream-related operations, when using  custom compiler and aot_module_simplified? 

cc @ezyang @anijain2305 @chauhang @penguinwu @zou3519 @ptrblck @msaroufim @yf225

Here is my test script. 
When I use aot_toy_backend backend, no stream related ops in gx graph.
What can we do to fix this？ Can you give me some guidance or advice on this issue.

```
import torch
import torch.nn as nn
class Layer(nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x):
        stream2 = torch.cuda.Stream()
        with torch.cuda.stream(stream2):
            z = x + 1
        y = x - 1
        return y + z

mm = Layer()
x=torch.randn([4]).cuda()

from torch._functorch.aot_autograd import aot_module_simplified
def toy_backend(gm, sample_inputs):
    return gm
def aot_toy_backend(gm, sample_inputs):
    return aot_module_simplified(gm, sample_inputs, fw_compiler=toy_backend)

mmc = torch.compile(mm, backend=aot_toy_backend)
yc= mmc(x)
```

### Versions

pytorch 2.3.0

**Comment by wbigat** [comment]
I would like to do some stream-related attempts on custom compiler.  I sincerely look forward to your reply on this issue. Thank you.   @yf225  @voznesenskym

**Comment by yf225** [comment]
@wbigat Would you like to talk about your motivating use case for having user-specified streams in the AOT joint graph?

There was a discussion on supporting user-specified streams in AOT joint graph and Inductor, and the general consensus is that it's a lot of work and the upside is unclear. (Please see the discussion notes starting with "Streams is hard, it doesn’t seem like we want to solve streams hard enough to actually solve them." in [Composability meeting notes](https://docs.google.com/document/d/1QTR3t3KdRu5JT1lvAuJLsPd3LruCfv0LecpsgR8eWhg/edit?usp=sharing).)

If you just want to use streams to parallelize compute, then *not* specifying user stream and letting Inductor automatically find out the right parallelization plan is our plan of record.

If you want to use streams to parallelize several communication ops, or parallelize a mix of compute and communication ops, we'd be curious to hear your use case.

**Comment by wbigat** [comment]
@yf225  Thank you very much for your reply.
I understand that if I use Inducer, then letting Inducer decide how to run multiple operators in parallel is what you are planning. This issue mainly  focuses on how to deal with this problem in the custom compiler. As you say, parallelism between multiple compute ops, and parallelism between compute and communication are the points we want to try.

As an example, in a Dense-MoE Hybrid architecture, we can overlap all-to-all with Attention/MLP computations.
And for the overlap of computing and communication mentioned above, we tried it in the eager mode. When this stream-based parallelism is implemented, better inference performance is indeed achieved.

So, I want to be able to get the stream ops written in the python script when using custom compiler.

The current obstacle is when I registered a custom compiler, like the previous example ```aot_toy_backend```, after the ```aot_module_simplified``` process, stream related operations have been stripped,  and can not be displayed in the fx graph obtained by the custom compiler.

This is the fx graph to custom compiler:
```
def forward(self, arg0_1):
    add = torch.ops.aten.add.Tensor(arg0_1, 1)
    sub = torch.ops.aten.sub.Tensor(arg0_1, 1);  arg0_1 = None
    add_1 = torch.ops.aten.add.Tensor(sub, add);  sub = add = None
    return add_1
```

**Comment by yf225** [comment]
@wbigat To understand more, are you using NVIDIA or AMD gpus as underlying hardware, just with a different custom compiler?

**Comment by wbigat** [comment]
> @wbigat To understand more, are you using NVIDIA or AMD gpus as underlying hardware, just with a different custom compiler?

When I run with eager mode, I use some NV GPUs. When I run with custom compiler, I use some NPUs.

**Comment by yf225** [comment]
> When I run with custom compiler, I use some NPUs.

I think for this case, if it's only for inference, you can lower to custom compiler from Dynamo graph, which will still have stream information.

For training, we need AOTAutograd to support streams as well which I think will be a lot more work. 

Fundamentally we could preserve user-specified streams through the lowering stack, but this would require considerable amount of investment, and we'd like to build up enough momentum on this before investing in it.

**Comment by wbigat** [comment]
> > When I run with custom compiler, I use some NPUs.
> 
> I think for this case, if it's only for inference, you can lower to custom compiler from Dynamo graph, which will still have stream information.
> 
> For training, we need AOTAutograd to support streams as well which I think will be a lot more work.
> 
> Fundamentally we could preserve user-specified streams through the lowering stack, but this would require considerable amount of investment, and we'd like to build up enough momentum on this before investing in it.

Using for inference is my main purpose. When I define a custom compiler as above, there is not any stream info in it. Just like the testcase in issue.

Sorry, I don't understand how you mean to lower stream to custom compiler? Can you provide more detailed explanations or some examples? Thank you very much.

**Comment by wbigat** [comment]
@yf225    Would you mind giving me some more guidance about how to lower to custom compiler from Dynamo graph?

**Comment by wbigat** [comment]
> If you want to use streams to parallelize several communication ops, or parallelize a mix of compute and communication ops, we'd be curious to hear your use case.

In a Dense-MoE Hybrid architecture, we can overlap all-to-all with Attention/MLP computations. This is the actual case of communication and computation parallelism that I want to implement.  Would you mind taking a look?

**Comment by wbigat** [comment]
> you can always graph-break and carve out the user-stream region to be executed in eager mode

I understand what you're saying, and I definitely agree that this would be a solution to parallelism.

But in my actual case, the architecture to be run in parallel is in a layer that will be repeated many times. So when I try to carve out the user-stream region to be executed in eager mode, there are lots of scattered subgraphs and operators.

There is a lot of header overhead in each graph execution in this way, which does not achieve the performance gains I want. I still want to try to get all the stream operators in full graph. I sincerely look forward to your reply. @yf225

**Comment by yf225** [comment]
@wbigat Even without user-specified stream, PyTorch all-to-all collective already runs on a separate stream. Assuming the AOTAutograd graph preserves this behavior, is it sufficient for your case?

You can observe this behavior by adding more compute ops to this unit test: https://github.com/pytorch/pytorch/blob/b4b64f76e5561902777e1a39b2020417a3e3ea31/test/distributed/test_c10d_functional_native.py#L718, run `torch.compile` with `aot_eager` backend in this unit test and check the generated AOTAutograd graphs (with `TORCH_LOGS="aot_graphs,aot_joint_graph"`).

**Comment by wbigat** [comment]
> @wbigat Even without user-specified stream, PyTorch all-to-all collective already runs on a separate stream. Assuming the AOTAutograd graph preserves this behavior, is it sufficient for your case?
> 
> You can observe this behavior by adding more compute ops to this unit test:
> 
> https://github.com/pytorch/pytorch/blob/b4b64f76e5561902777e1a39b2020417a3e3ea31/test/distributed/test_c10d_functional_native.py#L718
> 
> , run `torch.compile` with `aot_eager` backend in this unit test (you can check the generated AOTAutograd graphs with `TORCH_LOGS="aot_graphs,aot_joint_graph"`). If you enable PyTorch profiler tracing, you will see that the all-to-all collective is already run on a separate stream, and all we need to do would be to move the wait op to a later point in graph to achieve compute-comm overlap.

Yes, this  is a parallel strategy.  But I want to achieve more refined stream control. For example, running multiple computations and communication ops in parallel on two independent streams, with no data dependencies between the two streams.
 If I don't use stream at this point, relying solely on communication ops is not enough to achieve such parallelism.

**Comment by yf225** [comment]
Got it, in this case I believe lowering from Dynamo graph makes sense. This is an example:
```python
import torch

def my_custom_backend(gm, example_inputs):
    print(f"gm: {gm}")
    return gm.forward

def func(x):
    s = torch.cuda.Stream()
    A = torch.empty((100, 100), device="cuda").normal_(0.0, 1.0)
    s.wait_stream(torch.cuda.default_stream("cuda"))
    with torch.cuda.stream(s):
        B = torch.sum(A)
    A.record_stream(s)
    return A + x

torch.compile(func, backend=my_custom_backend)(torch.randn(1, device="cuda"))

"""
this prints:

def forward(self, L_A_ : torch.Tensor, L_x_ : torch.Tensor):
    l_a_ = L_A_
    l_x_ = L_x_
    stream = torch.cuda.streams.Stream(stream_id = 35, device_index = 0, device_type = 1)
    current_stream = torch.cuda.current_stream(None)
    set_stream = torch.cuda.set_stream(stream)
    B = torch.sum(l_a_)
    set_stream_1 = torch.cuda.set_stream(current_stream);  current_stream = None
    record_stream = l_a_.record_stream(stream);  stream = None
    add = l_a_ + l_x_;  l_a_ = l_x_ = None
    return (add,)
"""
```
Notice that all the stream operations are still in the graph, and you can leverage them in the custom compiler to have fine-grained stream control.

**Comment by wbigat** [comment]
> Got it, in this case I believe lowering from Dynamo graph makes sense. This is an example:
> 
> ```python
> import torch
> 
> def my_custom_backend(gm, example_inputs):
>     print(f"gm: {gm}")
>     return gm.forward
> 
> def func(x):
>     s = torch.cuda.Stream()
>     A = torch.empty((100, 100), device="cuda").normal_(0.0, 1.0)
>     s.wait_stream(torch.cuda.default_stream("cuda"))
>     with torch.cuda.stream(s):
>         B = torch.sum(A)
>     A.record_stream(s)
>     return A + x
> 
> torch.compile(func, backend=my_custom_backend)(torch.randn(1, device="cuda"))
> 
> """
> this prints:
> 
> def forward(self, L_A_ : torch.Tensor, L_x_ : torch.Tensor):
>     l_a_ = L_A_
>     l_x_ = L_x_
>     stream = torch.cuda.streams.Stream(stream_id = 35, device_index = 0, device_type = 1)
>     current_stream = torch.cuda.current_stream(None)
>     set_stream = torch.cuda.set_stream(stream)
>     B = torch.sum(l_a_)
>     set_stream_1 = torch.cuda.set_stream(current_stream);  current_stream = None
>     record_stream = l_a_.record_stream(stream);  stream = None
>     add = l_a_ + l_x_;  l_a_ = l_x_ = None
>     return (add,)
> """
> ```
> 
> Notice that all the stream operations are still in the graph, and you can leverage them in the custom compiler to have fine-grained stream control.

Yes, defining a function directly as a compiler like this can obtain stream related operators, and I have also conducted relevant tests. 

Besides, I have something that I don't quite understand. The compilers I previously defined all used aot_module_simplified, just like the issue example. Now that aot_module_simplified  is no longer used in my_custom_backend, does any additional work need to be done for the my_custom_backend ?

**Comment by yf225** [comment]
> does any additional work need to be done for the my_custom_backend ? Just like the inplace operators.

yes inplace operators is probably the biggest one, because it's normally handled in AOTAutograd. Now that we are not using AOTAutograd here, the custom compiler will need to handle inplace ops.

> And if I only use it in inference, is aot_module_simplified no longer a necessary step?

yes it's generally not a strictly required step for inference, especially if it removes stream operators which is undesirable in this use case.
